Merge RD (#22)
stephenneale/seng371-group16: as-developers/rd.md

| Students    | A group of users that are registered to Courses and can see the content posted therein              |

### 1.4 Overview

## 2 Overall Description

### 2.1 Product Perspective

### 2.2 Product Features

### 2.3 User Classes and Characteristics

### 2.4 Operating Environment

### 2.5 Design and Implementation Constraints

### 2.6 Assumptions and Dependencies

#### 2.6.1 Assumptions

#### 2.6.2 Dependencies

## 3 System Features

### 3.1 Courses

#### 3.1.1 Description and Priority

There are two main account types for the courses section, students and instructors. A student account will be able to view and access course content for only courses that they are registered in. The homepage consists of a catalogue of courses that the student has registered in. The instructors account will only have access to courses that they own and will be able to control what the students have access to with-in their own course. They will also have a similar homepage which displays a catalogue of courses which they own. This feature is of **high priority**.
#### 3.1.2 Functional Requirements

> REQ-1-1: Each homepage must only show courses that a student is registered in or an instructor owns

Author: Just suggesting a little re-word for clarity
```suggestion
REQ-1-1: Students' homepages should show exactly the courses that they are registered in, and instructors' homepages should show exactly the courses that they administer
```

> REQ-1-2: Each course must display the course name, course number and the section that the student belongs to

Author: Oxford comma gang?

```suggestion
REQ-1-2: Each course must display the course name, course number, and the section that the student belongs to
```

that they administer. Users can interact with this in a standard web-app fashion (point with mouse
pointer and click, keyboard, etc), as a later-stage goal this interface should be accessible to users
with a wide variety of ability. In particular, it should play nice with screen readers, and shouldn't
communicate essential information using colour in a way that would cause difficulties for people with
colour blindness.

### 4.2 Hardware Interfaces

Given that this is primarily a software system, hardware details will be left to a cloud service vendor
such as Amazon AWS, Google Cloud, or Microsoft Azure. We only require that the service be performant (as
described in section 5 of this document), which does involve hardware considerations, but those are outside
the scope of this document.

### 4.3 Software Interfaces

The system will interface with a database to durably store information.

### 4.4 Communications Interfaces

The system will be accessible through the Internet, so that students and instructors can view their
courses from anywhere with an Internet connection.

## 5 Other Non-Functional Requirements

### 5.1 Performance Requirements

> The system needs to run at high speeds, 24/7, with low failure rate. Due to students needing a reliable system, it should be stress tested for 160+ hours a week and a capacity of 2500+ students at one time. System updates must happen live or nearly live as this ensures students will not need to worry about downtimes interrupting their studies.

Author: Some rewording for clarity

```suggestion
The system needs to run at high speeds, 24/7, with low failure rate. Since students need a reliable system, it should be stress tested for 160+ hours a week and a capacity of 2500+ students at one time. System updates must happen without appreciable downtime, to ensure that students' studies will not be adversely affected.
```

### 5.2 Security Requirements

> Account and data security are important as the system will be holding data of students such as names, student number, email and grades. The system will also be holding the class lectures and notes/assignments created by professors at the school which needs to protected.

Author: Some rewording for clarity

```suggestion
Account and data security are important as the system will store students' personal and private information (e.g. their names, student numbers, email addresses, and grades). The system will also store copyrighted and otherwise private information created by professors (notes, lectures, and assignments), and students (mainly assignment submissions).
```

submitted and the grades they received on their own work (students can't see each other's submissions
or grades). This feature is of **high priority**.

#### 3.2.2 Functional Requirements

REQ-2-1: Instructors may post notes (discrete bulletin style messages) to their course pages.

REQ-2-2: Instructors may add arbitrary file attachments to these notes that may be viewed and
downloaded by students

REQ-2-3: Students see all notes that instructors post on course pages in reverse chronological
order on the respective course page

REQ-2-4: Instructors may designate specific notes to accept assignment submissions. Students can upload
files to these assignment notes, which are in turn visible to the instructor. These submissions
are visible only to the student who uploaded them, and the instructor.

REQ-2-5: Instructors may assign grades (decimal number from 0.0 to 100.0) to each student's submission,
which the student can see in the place where they submitted their assignment.

REQ-2-6: Instructors may optionally set a due date for assignments, after which students cannot
submit anything further.

### 3.3 Grades

#### 3.3.1 Description and Priority

> This feature must allow the user to see the grade number for all the items that has been assigned a grade by the instructor. When the 'Grade' tab is clicked, all the graded items will be presneted using a spreadsheet style. Each item should have the following criterias: item name, grade, and weight achieved. In additon, a instructor user should have the permission to input a grade number for an item and system will automatically calculate the weight achieved, but a student user will only have the permission to view the items that have already been assigned a grade by the instructor. This feature is off **hight priority**.

Author: ```suggestion
This feature must allow the user to see the grade number for all the items that have been assigned a grade by the instructor. When the 'Grade' tab is clicked, all the graded items will be presented using a spreadsheet style. Each item should have the following information: item name, grade, and weight achieved. In addition, a instructor user should have permission to input a grade number for an item, and system will automatically calculate the weight achieved, but a student user will only have the permission to view the items that have already been assigned a grade by the instructor. This feature is of **high priority**.
```

#### 3.3.2 Functional Requirements

> REQ-3-1: Students see all the graded items' item name, grade and weight achieved on one page in chronological order when navigated to 'Grade' page. 

Author: ```suggestion
REQ-3-1: Students see all the graded items' item names, grades and weights achieved on one page in chronological order on the 'Grades' page. 
```

#### 2.3.2 Instructor Users
Instructor users will have the permission to modify all the objects. They will be able to upload files under the Course Content section and assign a starting and ending date to each object under the section. They will also be able to input grade numbers under the Grade section and assign a formula of how the weight achieved is automatically calculated. 

### 2.4 Operating Environment
- Hardware platform: The system can be accessed using common web browser apps on both PC's and portable devices. The web browser apps need to support basic HTML and JavaScript extensions.
- Operating system and version
 
### 2.5 Design and Implementation Constraints

_Hardware constraints_

Students generally have access to relatively recent hardware and modern browsers, so support for older browsers such as
Internet Explorer is not necessary. Since we will rely on cloud services for hosting, we will not consider hardware
requirements—this responsibility will be left to the cloud provider.

_Interaction with Third-Party Applications_

> The software may make use of third-party software libraries for development purposes as long as there is no concern of

Author: If you'd like to suggest a change, hit the suggestion button above the comment box...
```suggestion
The software may make use of third-party software libraries for development purposes when there is no concern of
```

> The software may make use of third-party software libraries for development purposes as long as there is no concern of

Author: If you'd like to suggest a change, you can use the suggestion button above the comment box…
```suggestion
The software may make use of third-party software libraries for development purposes when there is no concern of
```
